### PR TITLE
add logarithme mecanism

### PR DESCRIPTION
**Pull request by pbourdu**
close https://github.com/publicodes/publicodes/issues/608

### CODE REVIEW

**Comment by pbourdu** [comment]
@johangirod , j'ai tenté une implémentation (avec mon degré de compréhension / limité).
À dispo pour traiter les commentaires de review qu'il y aura

File: packages/core/src/mecanisms/logarithme.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    const explanation = parse(v, context);
+    return {    
+        explanation,
+        nodeKind: 'logarithme',
+    } as LogarithmeNode;
+}
+
+const evaluate: EvaluationFunction<'logarithme'> = function (node) {
+    const value = this.evaluateNode(node.explanation);
+    let nodeValue = Math.log(value.nodeValue as number);
```
Comment (EmileRolley) [inline packages/core/src/mecanisms/logarithme.ts:22]: remarque : idéalement on voudrait forcer le type de la `nodeValue` d'un `LogarithmeNode` pour être `Evaluation<number>`. Or, je n'ai pas l'impression que ce soit faisable en l'état, et je pense que si on souhaite rajouter la possibilité de paramétriser le type de `ASTNode` cela devra être fait dans une autre PR. 

Pour l'instant, il n'est pas possible d'avoir la garantie que la valeur sera effectivement un nombre. **Il faudrait à minima rajouter une vérification dans la fonction `evaluate` (reste à determiner le comportement attendu)** ou bien rajouter une vérification au moment du parsing, mais j'ai peur que ça nécessite de mettre en place une passe de typechecking, donc pour une prochaine PR également.

File: packages/core/test/mécanismes/logarithme.yaml
Comment (EmileRolley) [inline packages/core/test/mécanismes/logarithme.yaml:1]: suggestions : je pense qu'il y a la place pour rajouter quelques edge cases, permettant notamment de définir le comportement attendu pour lorsque les valeurs ne sont pas des nombres.

**Review by EmileRolley — COMMENTED** [review]
Merci @pbourdu pour la PR !

Je me permet de faire une première petite review. Dans l'ensemble ça me parait très bien. 

Il manque simplement à définir le comportement souhaiter lorsque la valeur sur laquelle le logarithme est appliqué n'est pas un nombre. Personnellement, je serai pour avoir une erreur de typage au parsing mais cela nécessitera sûrement un peu de travail de refacto au niveau de l'architecture du moteur. Qu'est-ce que tu en penses @johangirod ?

Si tu peux également rajouter une section expliquant le comportement du mécanisme et un exemple dans https://github.com/pbourdu/publicodes/blob/master/website/src/routes/docs/mecanismes/%2Bpage.svelte.md ce serait top également :+1:

File: packages/core/src/mecanisms/logarithme.ts (hunk 1 above)
Comment (johangirod) [reply packages/core/src/mecanisms/logarithme.ts:22]: Ça me va comme comportement. Par ailleurs, il faut ajouter un cas pour vérifier si la valeur n'est pas `undefined` (manquante) ou `null` (non applicable).

Si la valeur est un nombre, on appelle le logarithme. Si la valeur est undefined ou null, on la retourne. Sinon, on lève une erreur d'évaluation avec `new PublicodesError`.

File: packages/core/test/mécanismes/logarithme.yaml
Comment (EmileRolley) [reply packages/core/test/mécanismes/logarithme.yaml:1]: Est-ce que l'on souhaiterait pas rajouter la possibilité de tester les cas sensés lever une erreur ? Je n'ai pas regardé mais a priori ça ne me semblerait pas très compliqué à rajouter non :thinking: ?

Comment (johangirod) [reply packages/core/test/mécanismes/logarithme.yaml:1]: Oui, totalement, ça fait un moment que j'y pense, ça prendrait 10 minutes. Je vais l'ajouter dans une prochaine PR.

File: packages/core/src/AST/index.ts [hunk 2]
```diff
@@ -171,6 +171,8 @@ export const traverseASTNode: TraverseFunction<NodeKind> = (fn, node) => {
 			return traverseTextNode(fn, node)
 		case 'condition':
 			return traverseConditionNode(fn, node)
+		case 'logarithme':
+			return traverseLogarithmNode(fn, node)
```
[the remark this answers is not in the document]
Comment (pbourdu) [reply packages/core/src/AST/index.ts:175]: Ok modifié

File: packages/core/src/inferNodeType.ts [hunk 3]
```diff
@@ -40,6 +40,7 @@ export default function inferNodesTypes(
 			case 'grille':
 			case 'taux progressif':
 			case 'inversion':
+			case 'logarithme':
 			case 'résoudre référence circulaire':
 				return { isNullable: false, type: 'number' }
```
[the remark this answers is not in the document]
Comment (pbourdu) [reply packages/core/src/inferNodeType.ts:45]: Pas pu utiliser exactement la même implémentation que arrondi car pas le même format de `explanation` 🤔.
J'ai déplacé et utilisé sensiblement la même implémentation que `arrondi`.

File: packages/core/src/mecanisms/logarithme.ts (hunk 1 above)
Comment (pbourdu) [reply packages/core/src/mecanisms/logarithme.ts:22]: Ok, j'ai ajouté une validation et la levée d'une erreur d'évaluation sinon.

File: packages/core/test/mécanismes/logarithme.yaml
Comment (pbourdu) [reply packages/core/test/mécanismes/logarithme.yaml:1]: Sur ce point je ne crois pas avoir d'action particulière à réaliser.

**Comment by pbourdu** [comment]
> Merci @pbourdu pour la PR !
> 
> Je me permet de faire une première petite review. Dans l'ensemble ça me parait très bien.
> 
> Il manque simplement à définir le comportement souhaiter lorsque la valeur sur laquelle le logarithme est appliqué n'est pas un nombre. Personnellement, je serai pour avoir une erreur de typage au parsing mais cela nécessitera sûrement un peu de travail de refacto au niveau de l'architecture du moteur. Qu'est-ce que tu en penses @johangirod ?
> 
> Si tu peux également rajouter une section expliquant le comportement du mécanisme et un exemple dans https://github.com/pbourdu/publicodes/blob/master/website/src/routes/docs/mecanismes/%2Bpage.svelte.md ce serait top également 👍

C'est fait.

J'ai traité vos commentaires merci @johangirod @EmileRolley !

**Review by johangirod — COMMENTED** [review]
Merci @pbourdu :pray: 
On y presque, j'ai juste ajouté deux petites suggestions. Enfin, pour que cette nouvelle fonctionnalité apparaisse automatiquement au changelog, tu peux suivre les instructions du commentaire dédié : 

https://github.com/publicodes/publicodes/pull/609#issuecomment-2599650173

Une fois cela fait, on est bon pour le merge :ok_hand:

File: packages/core/test/mécanismes/logarithme.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+mon résultat:
+  logarithme: 1
+
+ma variable: 12
+résultat:
+  titre: Le log de ma variable
+  logarithme: ma variable
+  arrondi: 2 décimales
```
[the remark this answers is not in the document]
Comment (pbourdu) [reply packages/core/test/mécanismes/logarithme.yaml:8]: Updated

File: website/src/routes/docs/mecanismes/+page.svelte.md [hunk 5]
```diff
@@ -675,3 +675,16 @@ revenu professionnel:
   résoudre la référence circulaire: oui
   unité: €/an
 ```
+
+## `logarithme`
+
+Calcule le logarithme.
```
[the remark this answers is not in the document]
Comment (pbourdu) [reply website/src/routes/docs/mecanismes/+page.svelte.md]: Updated

**Comment by pbourdu** [comment]
> Merci @pbourdu 🙏 On y presque, j'ai juste ajouté deux petites suggestions. Enfin, pour que cette nouvelle fonctionnalité apparaisse automatiquement au changelog, tu peux suivre les instructions du commentaire dédié :
> 
> [#609 (comment)](https://github.com/publicodes/publicodes/pull/609#issuecomment-2599650173)
> 
> Une fois cela fait, on est bon pour le merge 👌

Hello,
Je viens de faire les modifications demandées, et j'ai résolu des conflits avec la dernière version de `master`.
Cette fois-ci j'espère que ça va pouvoir être merge 🤞 .
Merci

**Comment by johangirod** [comment]
Je merge ! La fonctionnalité sera shippée dans la prochaine version `1.7.0`.
Merci d'avoir proposé la PR et d'avoir fait les corrections suite aux retours :blue_heart:
